clang/lib/Frontend/Rewrite/InclusionRewriter.cpp

```diff
@@ -332,12 +346,14 @@ void InclusionRewriter::CommentOutDirective(Lexer &DirectiveLex,
     // OutputContentUpTo() would not output anything anyway.
     return;
   }
-  OS << "#if 0 /* expanded by -frewrite-includes */" << MainEOL;
+  OS << "#if 0 /* " << getIncludedFileName(Inc)
```

Comment:
I DO think there is at least a good attempt that could be made to get a unique set of names for each (perhaps a stringify of the path, ala how `#line` does it?), but I think this is a good stand-alone feature here.

One thing I MIGHT suggest anyway though is replacing the '0' with `defined(__CLANG_REWRITTEN_INCLUDES)`, so that turning on the `#include` version would be trivial.

I might also suggest for certain lines, `defined(__CLANG_REWRITTEN_SYSTEM_INCLUDES)` such that only the system headers are used from the test machine.

So that is, a system header would be:

```#if defined(__CLANG_REWRITTEN_INCLUDES) || defined(__CLANG_REWRITTEN_SYSTEM_INCLUDES) /* vector expanded by -frewrite-includes */```

Reply:
TIL about `#pragma clang system_header` which turns a header into a system header, regardless of how it was found. But, by the time the rewriter sees the pragma, it has already emitted the `#include` and its wrapper `#if`. That makes the `__CLANG_REWRITTEN_SYSTEM_INCLUDES` not as reliable, and I didn't do that part (yet).
But, I did do the `defined(__CLANG_REWRITTEN_INCLUDES)` part, so you can get all-or-nothing.

Reply:
Hmm... ok.  The 'all or nothing' is fine.  As far as the system-includes, I wouldn't mind it if we were imperfect, as this is exclusively a debugging tool, thus any amount of ABI/API/etc stability doesn't really exist.  So I'm ok with a bit of a 'best effort' here (same on the enabling individual headers).  

we can always revise this as we determine needs.

BTW, I'm ok with this patch, but I am not the code owner here, so I'd like to give others a better chance to review this, particularly @MaskRay .

Reply:
Aha, `-MM` has the same problem, so I'm less concerned about that now. I'll add the toggle for "system" headers later today.